lib/active_merchant/billing/gateways/d_local.rb

```diff
@@ -250,6 +255,41 @@ def signature(post, timestamp)
       def post_data(action, parameters = {})
         parameters.to_json
       end
+
+      def xid_or_ds_trans_id(three_d_secure)
+        if three_d_secure[:version].to_f >= 2
+          { ds_transaction_id: three_d_secure[:ds_transaction_id] }
+        else
+          { xid: three_d_secure[:xid] }
+        end
+      end
+
+      def add_three_ds(post, options)
+        return unless three_d_secure = options[:three_d_secure]
+
+        post[:three_dsecure] = {
+          mpi: true,
+          three_dsecure_version: three_d_secure[:version],
+          cavv: three_d_secure[:cavv],
+          eci: three_d_secure[:eci],
+          enrollment_response: three_d_secure[:enrolled] == 'true' ? 'Y' : 'N',
+          authentication_response: three_d_secure[:authentication_response_status]
+        }.merge(xid_or_ds_trans_id(three_d_secure))
+      end
+
+      def validate_three_ds_params(three_ds)
```

Comment:
I will say we usually do not perform this kind of validation in AM adapters, allowing the gateway to respond with their own validation error. But I think there is no harm in including it.